tools/checklocks/facts.go

```diff
@@ -490,6 +501,62 @@ func (pc *passContext) exportLockGuardFacts(ts *ast.TypeSpec, ss *ast.StructType
 	}
 }
 
+func isMutexType(typeName string) bool {
+	return mutexRE.MatchString(typeName) || rwMutexRE.MatchString(typeName)
+}
+
+// inferLockGuardFacts guesses whether fields are guarded under a mutex hat.
+func (pc *passContext) inferLockGuardFacts(ts *ast.TypeSpec, ss *ast.StructType) {
+	structType := pc.pass.TypesInfo.TypeOf(ts.Name).Underlying().(*types.Struct)
+
+	// mutexHatOf keeps track of the closest mutex above a line number.
+	// If the field at the line number is a mutex, returns itself.
+	mutexHatOf := make(map[int]string)
+
+	fset := pc.pass.Fset
+	for i, field := range ss.Fields.List {
+		fieldObj := structType.Field(i)
+
+		if typeName := fieldObj.Type().String(); isMutexType(typeName) {
```

Comment:
We can probably do away with typeName here and inline in the call to isMutexType.